rust-symcrypt/src/gcm.rs

```diff
@@ -0,0 +1,377 @@
+    auth_data: &[u8],
+    data: &[u8],
+    tag: &[u8],
+) -> Result<(), SymCryptError> {
+    unsafe {
+        match symcrypt_sys::SymCryptGcmValidateParameters(
+            block_cipher,
+            nonce.len() as symcrypt_sys::SIZE_T,
+            auth_data.len() as symcrypt_sys::UINT64,
+            data.len() as symcrypt_sys::SIZE_T,
+            tag.len() as symcrypt_sys::SIZE_T,
+        ) {
+            symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(()),
+            err => Err(err.into()),
+        }
+    }
+}
+
+pub fn gcm_encrypt(
+    key: &[u8],
+    nonce: &[u8],
+    auth_data: Option<&[u8]>,
+    src: &[u8],
+    tag: &mut [u8], // This is an out parameter
+) -> Result<Vec<u8>, SymCryptError> {
+    unsafe {
+        let mut expanded_key = symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY::default();
+        expand_key(key, &mut expanded_key)?;
+
+        let (auth_data_ptr, auth_data_len) = auth_data.map_or_else(
+            || (std::ptr::null(), 0 as symcrypt_sys::SIZE_T),
+            |data| (data.as_ptr(), data.len() as symcrypt_sys::SIZE_T),
+        );
+
+        let mut dst = vec![0u8; src.len()];
+
+        symcrypt_sys::SymCryptGcmEncrypt(
+            &mut expanded_key,
+            nonce.as_ptr(),
+            nonce.len() as symcrypt_sys::SIZE_T,
+            auth_data_ptr,
+            auth_data_len as symcrypt_sys::SIZE_T,
+            src.as_ptr(),
+            dst.as_mut_ptr(),
+            src.len() as symcrypt_sys::SIZE_T,
+            tag.as_mut_ptr(),
+            tag.len() as symcrypt_sys::SIZE_T,
+        );
+        Ok(dst)
+    }
+}
+
+pub fn gcm_decrypt(
+    key: &[u8],
+    nonce: &[u8],
+    auth_data: Option<&[u8]>,
+    src: &[u8],
+    tag: &[u8],
+) -> Result<Vec<u8>, SymCryptError> {
+    unsafe {
+        let mut expanded_key = symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY::default();
+        expand_key(key, &mut expanded_key)?;
+
+        let (auth_data_ptr, auth_data_len) = auth_data.map_or_else(
+            || (std::ptr::null(), 0 as symcrypt_sys::SIZE_T),
+            |data| (data.as_ptr(), data.len() as symcrypt_sys::SIZE_T),
+        );
+
+        let mut dst = vec![0u8; src.len()];
+
+        match symcrypt_sys::SymCryptGcmDecrypt(
+            &mut expanded_key,
+            nonce.as_ptr(),
+            nonce.len() as symcrypt_sys::SIZE_T,
+            auth_data_ptr,
+            auth_data_len,
+            src.as_ptr(),
+            dst.as_mut_ptr(),
+            src.len() as symcrypt_sys::SIZE_T,
+            tag.as_ptr(),
+            tag.len() as symcrypt_sys::SIZE_T,
+        ) {
+            symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(dst),
+            err => Err(err.into()),
+        }
+    }
+}
+
+// SymCrypt requires pointer address to stay static through the scope, passing expand_key as an out
+// parameter to maintain static pointer address
+fn expand_key(key: &[u8], expanded_key: &mut symcrypt_sys::SYMCRYPT_GCM_EXPANDED_KEY) -> Result<(), SymCryptError> {
```

Comment:
Opted for out parameter rather than `Box<>` either way the code is internal so not a big deal but the out parameter is slightly more clear than box for this circumstance in my opinion.